psga/common/primer_autodetection.nf

```diff
@@ -0,0 +1,81 @@
+/*
+ * Run: primer-autodetection (use first read only for illumina)
+ */
+process primer_autodetection {
+  publishDir "${params.output_path}/primer_autodetection", mode: 'copy', overwrite: true, pattern: '{*_primer.txt,*_primer_data.csv,*_primer_detection.tsv,trimmomatic.out,*.bowtie2*}'
+
+  tag "${task.index} - ${file_1}"
+
+  input:
+    path file_1
+    val kit
+
+  output:
+    path "*_primer.txt", emit: ch_primer
+    path "*_primer_data.csv", emit: ch_primer_data
+    path "*_primer_detection.tsv", emit: ch_primer_coverage
+    path "trimmomatic.out"
+    path "*.bowtie2*"
+
+  shell:
+  '''
+  kit=!{kit}
+  file_1=!{file_1}
+  sample_id=$( echo ${file_1} | cut -d '.' -f1 )
+  trimmed_sample="trimmed_sample.fastq"
+  crop="!{params.primer_minimum_length}"
+
+  # trim sample reads so that they are long as the shortest primer
+  java -jar /opt/trimmomatic/trimmomatic.jar SE -phred33 ${file_1} ${trimmed_sample} CROP:${crop} 2> trimmomatic.out
+
+  # iterate over the available primer scheme fasta.
+  # These fasta are generated by us and contain the sequences for each primer scheme name/version
+  for primer_fasta in $(find /primer_schemes -iname SARS-CoV-2.scheme.fasta); do
+      # extract the primer (e.g. ARTIC_V3)
+      primer="$(echo "${primer_fasta}" | cut -d/ -f3,5 | sed 's#/#_#')"
+
+      # build the index for the primer fasta
+      bowtie2-build ${primer_fasta} ${primer}.ref_idx &> ${primer}.bowtie2-build
+
+      # align the trimmed sample to the primer scheme fasta
+      bowtie2 --no-unal -x ${primer}.ref_idx -U ${trimmed_sample} 2> ${primer}.bowtie2 | samtools sort -o ${primer}.bam
```

Comment:
If it is matching, we could do it more efficiently for example by sorting the primers and navigating them like a tree. It seems like this will have to deal with every single primer in turn?

Reply:
Again, I don't think this is inefficient. It's actually rather fast and a standard approach for doing an alignment rather than implementing this with an ad-hoc script.